File: frontend/src/services/organizationService.ts
import { apiClient } from '@/lib/api';
import { ensureResponse, withAuth } from '@/services/base/BaseService';

// Department types
export interface Department {
  id: string;
  name: string;
  description: string | null;
  parent_id: string | null;
  manager_id: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface DepartmentWithChildren extends Department {
  children: DepartmentWithChildren[];
  teams_count: number;
}

export interface DepartmentCreate {
  name: string;
  description?: string | null;
  parent_id?: string | null;
  manager_id?: string | null;
}

export interface DepartmentUpdate {
  name?: string;
  description?: string | null;
  parent_id?: string | null;
  manager_id?: string | null;
  is_active?: boolean;
}

// Team types
export interface Team {
  id: string;
  name: string;
  description: string | null;
  department_id: string | null;
  lead_id: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface TeamWithMembers extends Team {
  members: TeamMemberWithUser[];
  department_name: string | null;
}

export interface TeamCreate {
  name: string;
  description?: string | null;
  department_id?: string | null;
  lead_id?: string | null;
}

export interface TeamUpdate {
  name?: string;
  description?: string | null;
  department_id?: string | null;
  lead_id?: string | null;
  is_active?: boolean;
}

// Team Member types
export interface TeamMember {
  id: string;
  team_id: string;
  user_id: string;
  role: string;
  joined_at: string;
}

export interface TeamMemberWithUser extends TeamMember {
  user_email: string;
  user_name: string;
  user_display_name: string | null;
}

export interface TeamMemberCreate {
  user_id: string;
  role?: string;
}

export interface TeamMemberUpdate {
  role?: string;
}

// Departments
async function getDepartments(): Promise<Department[]> {
  return withAuth(async () => {
    const response = await apiClient.get<Department[]>('/organization/departments');
    return ensureResponse(response, 'Invalid response from server');
  });
}

async function getDepartmentTree(): Promise<DepartmentWithChildren[]> {
  return withAuth(async () => {
    const response = await apiClient.get<DepartmentWithChildren[]>('/organization/departments/tree');
    return ensureResponse(response, 'Invalid response from server');
  });
}

async function createDepartment(data: DepartmentCreate): Promise<Department> {
  return withAuth(async () => {
    const response = await apiClient.post<Department>('/organization/departments', data);
    return ensureResponse(response, 'Invalid response from server');
  });
}

async function updateDepartment(departmentId: string, data: DepartmentUpdate): Promise<Department> {
  return withAuth(async () => {
    const response = await apiClient.patch<Department>(`/organization/departments/${departmentId}`, data);
    return ensureResponse(response, 'Invalid response from server');
  });
}

async function deleteDepartment(departmentId: string): Promise<void> {
  return withAuth(async () => {
    await apiClient.delete(`/organization/departments/${departmentId}`);
  });
}

// Teams
async function getTeams(departmentId?: string): Promise<TeamWithMembers[]> {
  return withAuth(async () => {
    const query = departmentId ? `?department_id=${departmentId}` : '';
    const response = await apiClient.get<TeamWithMembers[]>(`/organization/teams${query}`);
    return ensureResponse(response, 'Invalid response from server');
  });
}

async function getTeam(teamId: string): Promise<TeamWithMembers> {
  return withAuth(async () => {
    const response = await apiClient.get<TeamWithMembers>(`/organization/teams/${teamId}`);
    return ensureResponse(response, 'Invalid response from server');
  });
}

async function createTeam(data: TeamCreate): Promise<Team> {
  return withAuth(async () => {
    const response = await apiClient.post<Team>('/organization/teams', data);
    return ensureResponse(response, 'Invalid response from server');
  });
}

async function updateTeam(teamId: string, data: TeamUpdate): Promise<Team> {
  return withAuth(async () => {
    const response = await apiClient.patch<Team>(`/organization/teams/${teamId}`, data);
    return ensureResponse(response, 'Invalid response from server');
  });
}

async function deleteTeam(teamId: string): Promise<void> {
  return withAuth(async () => {
    await apiClient.delete(`/organization/teams/${teamId}`);
  });
}

// Team Members
async function addTeamMember(teamId: string, data: TeamMemberCreate): Promise<TeamMember> {
  return withAuth(async () => {
    const response = await apiClient.post<TeamMember>(`/organization/teams/${teamId}/members`, data);
    return ensureResponse(response, 'Invalid response from server');
  });
}

async function removeTeamMember(teamId: string, userId: string): Promise<void> {
  return withAuth(async () => {
    await apiClient.delete(`/organization/teams/${teamId}/members/${userId}`);
  });
}

async function getUserTeams(userId: string): Promise<Team[]> {
  return withAuth(async () => {
    const response = await apiClient.get<Team[]>(`/organization/users/${userId}/teams`);
    return ensureResponse(response, 'Invalid response from server');
  });
}

export const organizationService = {
  getDepartments,
  getDepartmentTree,
  createDepartment,
  updateDepartment,
  deleteDepartment,
  getTeams,
  getTeam,
  createTeam,
  updateTeam,
  deleteTeam,
  addTeamMember,
  removeTeamMember,
  getUserTeams,
};

File: backend/app/services/organization_service.py
from datetime import datetime, timezone
from uuid import UUID

from sqlalchemy import select, delete, func
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.models.db_models.organization import Department, Team, TeamMember
from app.models.db_models.user import User
from app.models.schemas.organization import (
    DepartmentCreate,
    DepartmentUpdate,
    TeamCreate,
    TeamUpdate,
    TeamMemberCreate,
    TeamMemberUpdate,
)


class OrganizationService:
    def __init__(self, db: AsyncSession):
        self.db = db

    # Department methods
    async def get_departments(self) -> list[Department]:
        result = await self.db.execute(
            select(Department)
            .options(selectinload(Department.teams))
            .order_by(Department.name)
        )
        return list(result.scalars().all())

    async def get_department(self, department_id: UUID) -> Department | None:
        result = await self.db.execute(
            select(Department)
            .options(selectinload(Department.teams))
            .filter(Department.id == department_id)
        )
        return result.scalar_one_or_none()

    async def create_department(self, data: DepartmentCreate) -> Department:
        department = Department(
            name=data.name,
            description=data.description,
            parent_id=data.parent_id,
            manager_id=data.manager_id,
        )
        self.db.add(department)
        await self.db.commit()
        await self.db.refresh(department)
        return department

    async def update_department(
        self, department_id: UUID, data: DepartmentUpdate
    ) -> Department | None:
        result = await self.db.execute(
            select(Department).filter(Department.id == department_id)
        )
        department = result.scalar_one_or_none()
        if not department:
            return None

        if data.name is not None:
            department.name = data.name
        if data.description is not None:
            department.description = data.description
        if data.parent_id is not None:
            department.parent_id = data.parent_id
        if data.manager_id is not None:
            department.manager_id = data.manager_id
        if data.is_active is not None:
            department.is_active = data.is_active

        department.updated_at = datetime.now(timezone.utc)
        await self.db.commit()
        await self.db.refresh(department)
        return department

    async def delete_department(self, department_id: UUID) -> bool:
        result = await self.db.execute(
            select(Department).filter(Department.id == department_id)
        )
        department = result.scalar_one_or_none()
        if not department:
            return False

        await self.db.delete(department)
        await self.db.commit()
        return True

    async def get_department_tree(self) -> list[dict]:
        """Get department hierarchy as a tree structure."""
        result = await self.db.execute(
            select(Department).order_by(Department.name)
        )
        departments = list(result.scalars().all())

        # Get team counts
        team_counts = {}
        for dept in departments:
            count_result = await self.db.execute(
                select(func.count()).select_from(Team).filter(
                    Team.department_id == dept.id,
                    Team.is_active == True,
                )
            )
            team_counts[str(dept.id)] = count_result.scalar() or 0

        # Build tree
        dept_map = {str(d.id): d for d in departments}
        root_depts = []

        for dept in departments:
            dept_dict = {
                "id": str(dept.id),
                "name": dept.name,
                "description": dept.description,
                "parent_id": str(dept.parent_id) if dept.parent_id else None,
                "manager_id": str(dept.manager_id) if dept.manager_id else None,
                "is_active": dept.is_active,
                "teams_count": team_counts.get(str(dept.id), 0),
                "children": [],
            }

            if dept.parent_id and str(dept.parent_id) in dept_map:
                pass  # Will be added as child
            else:
                root_depts.append(dept_dict)

        # Add children to parents
        for dept in departments:
            if dept.parent_id and str(dept.parent_id) in dept_map:
                for root in root_depts:
                    self._add_child_to_tree(root, dept, team_counts)

        return root_depts

    def _add_child_to_tree(
        self, tree: dict, dept: Department, team_counts: dict
    ) -> bool:
        if tree["id"] == str(dept.parent_id):
            tree["children"].append({
                "id": str(dept.id),
                "name": dept.name,
                "description": dept.description,
                "parent_id": str(dept.parent_id) if dept.parent_id else None,
                "manager_id": str(dept.manager_id) if dept.manager_id else None,
                "is_active": dept.is_active,
                "teams_count": team_counts.get(str(dept.id), 0),
                "children": [],
            })
            return True

        for child in tree.get("children", []):
            if self._add_child_to_tree(child, dept, team_counts):
                return True

        return False

    # Team methods
    async def get_teams(self, department_id: UUID | None = None) -> list[Team]:
        query = select(Team).options(
            selectinload(Team.members),
            selectinload(Team.department),
        )
        if department_id:
            query = query.filter(Team.department_id == department_id)
        query = query.order_by(Team.name)
        result = await self.db.execute(query)
        return list(result.scalars().all())

    async def get_team(self, team_id: UUID) -> Team | None:
        result = await self.db.execute(
            select(Team)
            .options(
                selectinload(Team.members),
                selectinload(Team.department),
            )
            .filter(Team.id == team_id)
        )
        return result.scalar_one_or_none()

    async def create_team(self, data: TeamCreate) -> Team:
        team = Team(
            name=data.name,
            description=data.description,
            department_id=data.department_id,
            lead_id=data.lead_id,
        )
        self.db.add(team)
        await self.db.commit()
        await self.db.refresh(team)
        return team

    async def update_team(self, team_id: UUID, data: TeamUpdate) -> Team | None:
        result = await self.db.execute(select(Team).filter(Team.id == team_id))
        team = result.scalar_one_or_none()
        if not team:
            return None

        if data.name is not None:
            team.name = data.name
        if data.description is not None:
            team.description = data.description
        if data.department_id is not None:
            team.department_id = data.department_id
        if data.lead_id is not None:
            team.lead_id = data.lead_id
        if data.is_active is not None:
            team.is_active = data.is_active

        team.updated_at = datetime.now(timezone.utc)
        await self.db.commit()
        await self.db.refresh(team)
        return team

    async def delete_team(self, team_id: UUID) -> bool:
        result = await self.db.execute(select(Team).filter(Team.id == team_id))
        team = result.scalar_one_or_none()
        if not team:
            return False

        await self.db.delete(team)
        await self.db.commit()
        return True

    # Team member methods
    async def add_team_member(
        self, team_id: UUID, data: TeamMemberCreate
    ) -> TeamMember:
        member = TeamMember(
            team_id=team_id,
            user_id=data.user_id,
            role=data.role,
        )
        self.db.add(member)
        await self.db.commit()
        await self.db.refresh(member)
        return member

    async def update_team_member(
        self, team_id: UUID, user_id: UUID, data: TeamMemberUpdate
    ) -> TeamMember | None:
        result = await self.db.execute(
            select(TeamMember).filter(
                TeamMember.team_id == team_id,
                TeamMember.user_id == user_id,
            )
        )
        member = result.scalar_one_or_none()
        if not member:
            return None

        if data.role is not None:
            member.role = data.role

        await self.db.commit()
        await self.db.refresh(member)
        return member

    async def remove_team_member(self, team_id: UUID, user_id: UUID) -> bool:
        result = await self.db.execute(
            delete(TeamMember).filter(
                TeamMember.team_id == team_id,
                TeamMember.user_id == user_id,
            )
        )
        await self.db.commit()
        return result.rowcount > 0

    async def get_user_teams(self, user_id: UUID) -> list[Team]:
        result = await self.db.execute(
            select(Team)
            .join(TeamMember)
            .filter(TeamMember.user_id == user_id)
            .options(selectinload(Team.department))
        )
        return list(result.scalars().all())
